File: rust/src/problem_solving/mcd_strings/reverse_string.rs
#[allow(dead_code)]
pub fn reverse_string(input: &str) -> String {
  let mut reversed = String::new();
  for c in input.chars().rev() {
    reversed.push(c);
  }
  reversed
}

#[allow(dead_code)]
pub fn reverse_recursive(input: &str) -> String {
  reverses_recursion_helper(input.to_string())
}

#[allow(dead_code)]
fn reverses_recursion_helper(mut s: String) -> String {
  if s.is_empty() {
    return s;
  }
  let removed_char = s.remove(0);
  let mut s = reverses_recursion_helper(s);
  s.push(removed_char);
  s
}

#[allow(dead_code)]
fn reverse_str_rec(s: &str) -> String {
  if s.len() == 0 {
    return String::from("");
  }
  let head = &s[0..1];
  let tail = &s[1..];
  let res = reverse_str_rec(tail);
  format!("{}{}", res, head)
}

#[cfg(test)]
mod tests {
  use super::*;

  #[test]
  fn it_works_to_reverse_string() {
    assert_eq!(reverse_string("abc"), "cba");
  }

  #[test]
  fn it_works_to_reverse_string_recursive() {
    assert_eq!(reverse_recursive("abc"), "cba");
  }

  #[test]
  fn it_works_to_reverse_string_rec() {
    assert_eq!(reverse_str_rec("abc"), "cba");
  }
}

File: just-algos-and-ds/algos/ts/src/algorithms/recursion/reverse_string.ts
export function reverse(input: string): string {
  if (input.length === 0) {
    return ""
  }
  const head = input.charAt(0)
  const tail = input.slice(1)
  // all the chars in the string except the first and add it to the end!
  return reverse(tail) + head
}

export function reverse2(input: string): string {
  if (input.length === 0) {
    return ""
  }
  return input[input.length - 1] + reverse2(input.slice(0, input.length - 1))
}
